File: src/ekore/anomalous_dimensions/polarized/space_like/as1.py
"""The |LO| Altarelli-Parisi splitting kernels."""

import numba as nb
import numpy as np

from eko import constants

from ....harmonics import cache as c
from ...unpolarized.space_like.as1 import gamma_ns


@nb.njit(cache=True)
def gamma_qg(N, nf):
    r"""Compute the |LO| polarized quark-gluon anomalous dimension.

    Implements :eqref:`A.1` from :cite:`Gluck:1995yr`.

    Parameters
    ----------
    N : complex
      Mellin moment
    nf : int
      Number of active flavors

    Returns
    -------
    complex
      |LO| polarized quark-gluon anomalous dimension :math:`\gamma_{qg}^{(0)}(N)`
    """
    gamma = -(N - 1) / N / (N + 1)
    result = 2.0 * constants.TR * 2.0 * nf * gamma
    return result


@nb.njit(cache=True)
def gamma_gq(N):
    r"""Compute the |LO| polarized gluon-quark anomalous dimension.

    Implements :eqref:`A.1` from :cite:`Gluck:1995yr`.

    Parameters
    ----------
    N : complex
      Mellin moment

    Returns
    -------
    complex
      |LO| gluon-quark anomalous dimension :math:`\gamma_{gq}^{(0)}(N)`
    """
    gamma = -(N + 2) / N / (N + 1)
    result = 2.0 * constants.CF * gamma
    return result


@nb.njit(cache=True)
def gamma_gg(N, cache, nf):
    r"""Compute the |LO| polarized gluon-gluon anomalous dimension.

    Implements :eqref:`A.1` from :cite:`Gluck:1995yr`.

    Parameters
    ----------
    N : complex
      Mellin moment
    cache: numpy.ndarray
        Harmonic sum cache
    nf : int
      Number of active flavors

    Returns
    -------
    complex
      |LO| gluon-gluon anomalous dimension :math:`\gamma_{gg}^{(0)}(N)`
    """
    S1 = c.get(c.S1, cache, N)
    gamma = -S1 + 2 / N / (N + 1)
    result = constants.CA * (-4.0 * gamma - 11.0 / 3.0) + 4.0 / 3.0 * constants.TR * nf
    return result


@nb.njit(cache=True)
def gamma_singlet(N, cache, nf):
    r"""Compute the |LO| polarized singlet anomalous dimension matrix.

      .. math::
          \gamma_S^{(0)} = \left(\begin{array}{cc}
            \gamma_{qq}^{(0)} & \gamma_{qg}^{(0)}\\
            \gamma_{gq}^{(0)} & \gamma_{gg}^{(0)}
          \end{array}\right)

    Parameters
    ----------
    N : complex
      Mellin moment
    cache: numpy.ndarray
        Harmonic sum cache
    nf : int
      Number of active flavors

    Returns
    -------
    numpy.ndarray
      |LO| singlet anomalous dimension matrix :math:`\gamma_{S}^{(0)}(N)`
    """
    gamma_qq = gamma_ns(N, cache)
    gamma_S_0 = np.array(
        [[gamma_qq, gamma_qg(N, nf)], [gamma_gq(N), gamma_gg(N, cache, nf)]],
        np.complex128,
    )
    return gamma_S_0

File: crates/ekore/src/anomalous_dimensions/polarized/spacelike/as1.rs
//! |LO| |QCD|.

use num::complex::Complex;

use super::super::super::unpolarized::spacelike::as1::gamma_ns as unpol;
use crate::constants::{CA, CF, TR};
use crate::harmonics::cache::{Cache, K};

/// Compute the non-singlet anomalous dimension.
///
/// Identical to the unpolarized counterpart.
pub fn gamma_ns(c: &mut Cache, nf: u8) -> Complex<f64> {
    unpol(c, nf)
}

/// Compute the quark-gluon anomalous dimension.
///
/// Implements Eq. (A.1) of [\[Gluck:1995yr\]][crate::bib::Gluck1995yr].
pub fn gamma_qg(c: &mut Cache, nf: u8) -> Complex<f64> {
    let N = c.n();
    let gamma = -(N - 1.) / N / (N + 1.);
    2.0 * TR * 2.0 * (nf as f64) * gamma
}

/// Compute the gluon-quark anomalous dimension.
///
/// Implements Eq. (A.1) of [\[Gluck:1995yr\]][crate::bib::Gluck1995yr].
pub fn gamma_gq(c: &mut Cache, _nf: u8) -> Complex<f64> {
    let N = c.n();
    let gamma = -(N + 2.) / N / (N + 1.);
    2.0 * CF * gamma
}

/// Compute the gluon-gluon anomalous dimension.
///
/// Implements Eq. (A.1) of [\[Gluck:1995yr\]][crate::bib::Gluck1995yr].
pub fn gamma_gg(c: &mut Cache, nf: u8) -> Complex<f64> {
    let N = c.n();
    let S1 = c.get(K::S1);
    let gamma = -S1 + 2. / N / (N + 1.);
    CA * (-4.0 * gamma - 11.0 / 3.0) + 4.0 / 3.0 * TR * (nf as f64)
}

/// Compute the singlet anomalous dimension matrix.
pub fn gamma_singlet(c: &mut Cache, nf: u8) -> [[Complex<f64>; 2]; 2] {
    let gamma_qq = gamma_ns(c, nf);
    [
        [gamma_qq, gamma_qg(c, nf)],
        [gamma_gq(c, nf), gamma_gg(c, nf)],
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::harmonics::cache::Cache;
    use crate::{assert_approx_eq_cmplx, cmplx};
    use num::complex::Complex;
    use num::Zero;
    const NF: u8 = 5;

    #[test]
    fn quark_momentum_conservation() {
        const N: Complex<f64> = cmplx!(2., 0.);
        let mut c = Cache::new(N);
        let me = gamma_ns(&mut c, NF) + gamma_gq(&mut c, NF);
        assert_approx_eq_cmplx!(f64, me, cmplx!((4. * CF) / 3., 0.), epsilon = 1e-12);
    }

    #[test]
    fn gluon_momentum_conservation() {
        const N: Complex<f64> = cmplx!(2., 0.);
        let mut c = Cache::new(N);
        let me = gamma_qg(&mut c, NF) + gamma_gg(&mut c, NF);
        assert_approx_eq_cmplx!(f64, me, cmplx!(3. + (NF as f64) / 3., 0.), epsilon = 1e-12);
    }

    #[test]
    fn qg_helicity_conservation() {
        const N: Complex<f64> = cmplx!(1., 0.);
        let mut c = Cache::new(N);
        let me = gamma_qg(&mut c, NF);
        assert_approx_eq_cmplx!(f64, me, Complex::zero(), epsilon = 1e-12);
    }
}
